feat(@astrojs/cloudflare): Add support for wasm module imports (#8542)
withastro/astro: packages/integrations/cloudflare/README.md

});
```

## WASM module imports

`wasmModuleImports: boolean`

> Cloudflare has native support for importing `.wasm` files [directly as ES modules](https://github.com/WebAssembly/esm-integration/tree/main/proposals/esm-integration). To enable importing them as modules in the cloudflare build and astro dev server, add `wasmModuleImports: true` to your config.

Reviewer: ```suggestion
default: `false`

Whether or not to import `.wasm` files [directly as ES modules](https://github.com/WebAssembly/esm-integration/tree/main/proposals/esm-integration). 

Add `wasmModuleImports: true` to `astro.config.mjs` to enable in both the Cloudflare build and the Astro dev server. 
```
This seems more in keeping with the option directly above, stating the default and firstly, defining the configuration setting itself.

```diff

> import {defineConfig} from "astro/config";

Reviewer: ```suggestion
// astro.config.mjs
import {defineConfig} from "astro/config";
```
We do add a file name whenever it's easy to do so, so that it's clear to someone just skimming code samples where this code is!

import cloudflare from '@astrojs/cloudflare';

export default defineConfig({
    adapter: cloudflare({
+       wasmModuleImports: true
    }),
	  output: 'server'
})
```

> Once enabled, you can import a web assembly module in astro with a `.wasm?module` import. The integration supports WASM module imports in both server and hybrid mode for a consistent development experience.

Reviewer: ```suggestion
Once enabled, you can import a web assembly module in Astro with a `.wasm?module` import. 

The following example shows... (importing ... then ...)
```
Just a reminder that proper names, when we just mention them in a sentence, should be capitalized.

Ideally, we prefer to introduce code samples with a sentence like, "The following example shows... " and then an explanation of exactly what the code is doing. This means people don't have to both figure out what this example is, and then also how the thing you're demonstrating also works. We help them by telling them before they see the example that this is what the code accomplishes.

Also, the fact that it works in both modes is great to include! But, it seems out of place here. I suggest moving this sentence below the code sample, since it's not actually related to this code example.

Author: Good points. I made the capitalization of `Wasm` consistent with that suggested at [https://webassembly.org/](https://webassembly.org/). And updated example introduction text to be more descriptive

```javascript
// pages/add/[a]/[b].js
import mod from '../util/add.wasm?module';

// instantiate ahead of time to share module
const addModule: any = new WebAssembly.Instance(mod);

export async function GET(context) {
  const a = Number.parseInt(context.params.a!);
	const b = Number.parseInt(context.params.b!);
  return new Response(`${a} + ${b} = ${addModule.exports.add(a, b)}`);
}
```

> 

Reviewer: ```suggestion
The Cloudflare integration supports WASM module imports in both `server` and `hybrid` mode for a consistent development experience.
```
This seems more appropriate as a "final note" rather than interrupting explaining usage.

Author: Reading this sentence over, it felt redundant. I would imagine that user expectation is that the feature works out of the box. 

Updated instead to give an idea of a more exciting potential use case for Wasm. I like the addition example because its simple and intuitive, but its not a practical use Wasm.

});
```

## Wasm module imports

`wasmModuleImports: boolean`

default: `false`

Whether or not to import `.wasm` files [directly as ES modules](https://github.com/WebAssembly/esm-integration/tree/main/proposals/esm-integration). 

Add `wasmModuleImports: true` to `astro.config.mjs` to enable in both the Cloudflare build and the Astro dev server. 

```diff
// astro.config.mjs
import {defineConfig} from "astro/config";
import cloudflare from '@astrojs/cloudflare';

export default defineConfig({
    adapter: cloudflare({
+       wasmModuleImports: true
    }),
	  output: 'server'
})
```

Once enabled, you can import a web assembly module in Astro with a `.wasm?module` import. 

> The following is a simple example of importing a Wasm module that then responds to requests by adding the request's number parameters together. A request to `/add/40/2` would return a response of `42`.

Reviewer: ```suggestion
The following is an example of importing a Wasm module that then responds to requests by adding the request's number parameters together. 
```
Nicely done, but even just this is sufficient! It's just about providing a little context!

Author: @sarah11918 thanks! Updated

```javascript
// pages/add/[a]/[b].js
import mod from '../util/add.wasm?module';

// instantiate ahead of time to share module
const addModule: any = new WebAssembly.Instance(mod);

export async function GET(context) {
  const a = Number.parseInt(context.params.a);
  const b = Number.parseInt(context.params.b);
  return new Response(`${addModule.exports.add(a, b)}`);
}
```

> While this example is trivial, Wasm can be used to accelerate computationally intensive operations which do not involve significant I/O. For example, embedding an image processing library.

Reviewer: ```suggestion
While this example is trivial, Wasm can be used to accelerate computationally intensive operations which do not involve significant I/O such as embedding an image processing library.
```
That just wasn't a full sentence on its own so reworded slightly!
